text/0000-module-unification.md

```diff
@@ -0,0 +1,822 @@
+4. Subtrees should be relocatable. If you move a directory to a new place in the tree, its internal structure should all still work.
+5. There should be no cognitive overhead when adding a new thing. The right way should be obvious and not impose unnecessary decisions.
+6. We need clean separation between the namespace of the user's own components, helpers, routes, etc and the framework's own type names ("component", "helper", etc) so that we can disambiguate them and add future ones.
+7. Ideally we will have a place to put tests and styles alongside corresponding components.
+8. Local relative lookup for components and helpers needs to work.
+9. Avoid the "titlebar problem", in which many files are all named "component.js" and you can't tell them apart in your editor.
+10. The resolver should be configured via declarative rules, not imperative
+    JavaScript. In addition to enforcing consistency, this allows addons to
+    augment the system with their own types in a predictable way.
+11. Module structures must be statically analyzable at build time to enable
+    efficiency optimizations.
+12. Module classifications must be extensible and allow for customizations by
+    apps, engines, and addons.
+
+> Note: Constraints > 9 were added based on discussions subsequent to the
+initial meeting.
+
+This proposal attempts to address these constraints with a single consistent
+approach to modules that will make Ember easier to use and learn _and_ improve
+the efficiency of Ember's resolver at run-time.
+
+# Detailed Design
+
+This proposal introduces a new top-level directory, `src`, and establishes
+conventions for organizing modules within it. Also proposed is a refactor of the
+Ember resolver to enable efficient and flexible resolutions based upon the new
+module conventions.
+
+The `src` directory will be used to contain the core ES modules within an Ember
+CLI project, whether that project contains an application, addon, or engine. To
+maintain backward compatibility, the `src` directory will be allowed to
+co-exist alongside existing `app` and/or `addon` directories, although these
+directories should eventually be deprecated.
+
+## Examples
+
+Let's start by taking a look at some examples of Ember projects organized
+according to the proposed conventions.
+
+### Example Application
+
+A simple blogging application could be structured as follows:
+```
+src
+├── data
+│   ├── models
+│   │   ├── comment
+│   │   │   ├── adapter.js
+│   │   │   ├── model.js
+│   │   │   └── serializer.js
+│   │   ├── post
+│   │   │   ├── adapter.js
+│   │   │   ├── model.js
+│   │   │   └── serializer.js
+│   │   └── author.js
+│   └── transforms
+│       └── date.js
+├── init
+│   ├── initializers
+│   │   └── i18n.js
+│   └── instance-initializers
+│       └── auth.js
+├── services
+│   └── auth.js
+├── ui
+│   ├── components
+│   │   ├── date-picker
+│   │   │   ├── component.js
```

Comment:
Clarification.

@rwjblue @rtm `component.js` is used 6 times in this application. I understand that this is a carry-over from pods and an improvement on classic, but it still feels like titlebar problem (9).

Would this be supported/encouraged as an option, or is it wrong?
```
│   ├── components
│   │   ├── date-picker
│   │   │   ├── date-picker.js
│   │   │   ├── date-picker.hbs
```
`component.js` and `date-picker.js` would be  considered 'main' (with one taking precedence in resolution) if there were multiple files of the same type.